File: src/CircleAI.Workflows/PacaBoards.cs
// PacaBoards.cs
//
// (3.3.0) Sprintboard surface ported from paca: rich JSON description,
// custom fields, story points, importance, parent/child relations,
// status columns with position-ordered workflow, drag-and-drop status
// transitions, sprints with lifecycle states, Scrumban swimlanes,
// per-view persistent configs (filters + sort + visible fields),
// tags, lazy-load pagination per column.

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace CircleAI.Workflows;

/// <summary>(3.3.0) Sprint lifecycle.</summary>
public enum SprintState { Planning, Active, Completed }

/// <summary>(3.3.0) Status column in the workflow.</summary>
public sealed record StatusColumn(
    string  Name,             // "todo" / "in_progress" / "in_review" / "done"
    string  Category,         // "open" / "in-flight" / "review" / "closed" / "cancelled" / "blocked"
    int     Position,
    bool    Collapsed);

/// <summary>(3.3.0) Sprint.</summary>
public sealed record PacaSprint(
    string         Id,
    string         ProjectId,
    string         Name,
    string         Goal,
    DateTimeOffset StartDate,
    DateTimeOffset EndDate,
    SprintState    State);

/// <summary>(3.3.0) Extra board-only metadata on top of <see cref="PacaTask"/>.</summary>
public sealed record TaskBoardMetadata(
    string             ProjectId,
    int                Number,
    int                StoryPoints,
    int                Importance,           // 0..5
    string?            AssigneeMemberId,
    string?            ReporterMemberId,
    int?               ParentTaskNumber,
    string?            SprintId,
    IReadOnlyList<string> Tags,
    IReadOnlyDictionary<string, string> CustomFields,
    int                PositionInColumn);

/// <summary>(3.3.0) A per-user / per-board "named view".</summary>
public sealed record BoardView(
    string                              Name,
    string?                             FilterTagsCsv,
    string?                             FilterAssignee,
    string?                             SortBy,           // "importance" / "story_points" / "newest"
    bool                                SortDescending,
    IReadOnlyList<string>               VisibleColumns,
    IReadOnlyList<string>               VisibleFields);

/// <summary>(3.3.0) Board service over a project. Sprints + columns + per-task metadata + views.</summary>
public sealed class PacaBoard
{
    private readonly InMemoryPacaStore _tasks;
    private readonly ConcurrentDictionary<string, StatusColumn>          _columns      = new();
    private readonly ConcurrentDictionary<string, PacaSprint>            _sprints      = new();
    private readonly ConcurrentDictionary<(string ProjectId, int Number), TaskBoardMetadata> _metadata = new();
    private readonly ConcurrentDictionary<string, BoardView>             _views        = new();
    private readonly Func<DateTimeOffset> _clock;

    public PacaBoard(InMemoryPacaStore tasks, Func<DateTimeOffset>? clock = null)
    {
        _tasks = tasks ?? throw new ArgumentNullException(nameof(tasks));
        _clock = clock ?? (() => DateTimeOffset.UtcNow);

        AddDefaultColumns();
    }

    private void AddDefaultColumns()
    {
        _columns["todo"]        = new StatusColumn("todo",        "open",      0, false);
        _columns["in_progress"] = new StatusColumn("in_progress", "in-flight", 1, false);
        _columns["in_review"]   = new StatusColumn("in_review",   "review",    2, false);
        _columns["done"]        = new StatusColumn("done",        "closed",    3, false);
        _columns["cancelled"]   = new StatusColumn("cancelled",   "cancelled", 4, false);
        _columns["blocked"]     = new StatusColumn("blocked",     "blocked",   5, true);
    }

    public IReadOnlyList<StatusColumn> Columns => _columns.Values.OrderBy(c => c.Position).ToList();

    public void AddColumn(StatusColumn col)
    {
        ArgumentNullException.ThrowIfNull(col);
        _columns[col.Name] = col;
    }

    public void CollapseColumn(string name, bool collapsed)
    {
        if (_columns.TryGetValue(name, out var col))
        {
            _columns[name] = col with { Collapsed = collapsed };
        }
    }

    /// <summary>(3.3.0) Move a task between status columns, updating its in-column position.</summary>
    public void MoveTask(string projectId, int number, string newStatus, int newPosition)
    {
        var task = _tasks.GetTaskByReference(projectId, $"{projectId}-{number}")
                   ?? _tasks.ListTasks(projectId).FirstOrDefault(t => t.Number == number);
        if (task is null) throw new InvalidOperationException("Task not found.");
        if (!_columns.ContainsKey(newStatus)) throw new ArgumentException($"Unknown status '{newStatus}'.", nameof(newStatus));

        _tasks.UpdateTask(task with { Status = newStatus });
        var meta = GetOrCreateMetadata(projectId, number) with { PositionInColumn = newPosition };
        _metadata[(projectId, number)] = meta;
    }

    /// <summary>(3.3.0) Attach board metadata to an existing task.</summary>
    public void SetTaskMetadata(TaskBoardMetadata metadata)
    {
        ArgumentNullException.ThrowIfNull(metadata);
        _metadata[(metadata.ProjectId, metadata.Number)] = metadata;
    }

    public TaskBoardMetadata? GetTaskMetadata(string projectId, int number)
        => _metadata.TryGetValue((projectId, number), out var meta) ? meta : null;

    /// <summary>(3.3.0) Paginated column read for lazy loading.</summary>
    public IReadOnlyList<PacaTask> TasksInColumn(string projectId, string status, int skip = 0, int take = 50)
    {
        var live = _tasks.ListTasks(projectId).Where(t => t.Status == status);
        return live
            .OrderBy(t => GetOrCreateMetadata(t.ProjectId, t.Number).PositionInColumn)
            .Skip(skip).Take(take).ToList();
    }

    /// <summary>(3.3.0) Tasks bucketed by sprint, useful for the Scrumban board.</summary>
    public IReadOnlyList<PacaTask> TasksInSprint(string sprintId)
    {
        return _metadata.Values
            .Where(m => m.SprintId == sprintId)
            .Select(m => _tasks.ListTasks(m.ProjectId).FirstOrDefault(t => t.Number == m.Number))
            .Where(t => t is not null)
            .Select(t => t!)
            .ToList();
    }

    /// <summary>(3.3.0) Create a sprint in Planning.</summary>
    public PacaSprint CreateSprint(string id, string projectId, string name, string goal, DateTimeOffset start, DateTimeOffset end)
    {
        var s = new PacaSprint(id, projectId, name, goal, start, end, SprintState.Planning);
        _sprints[id] = s;
        return s;
    }

    public PacaSprint? GetSprint(string id) => _sprints.TryGetValue(id, out var s) ? s : null;

    public PacaSprint StartSprint(string id)     => Transition(id, SprintState.Active);
    public PacaSprint CompleteSprint(string id)  => Transition(id, SprintState.Completed);

    private PacaSprint Transition(string id, SprintState to)
    {
        if (!_sprints.TryGetValue(id, out var sprint))
            throw new InvalidOperationException($"Sprint '{id}' not found.");
        var updated = sprint with { State = to };
        _sprints[id] = updated;
        return updated;
    }

    /// <summary>(3.3.0) Save a named view (filters + sort + visible fields).</summary>
    public void SaveView(BoardView view) => _views[view.Name] = view;

    public BoardView? GetView(string name) => _views.TryGetValue(name, out var v) ? v : null;

    public IReadOnlyList<BoardView> ListViews() => _views.Values.OrderBy(v => v.Name).ToList();

    private TaskBoardMetadata GetOrCreateMetadata(string projectId, int number)
    {
        return _metadata.GetOrAdd((projectId, number),
            _ => new TaskBoardMetadata(projectId, number, 0, 3, null, null, null, null,
                Array.Empty<string>(), new Dictionary<string, string>(), 0));
    }
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/workflows/PacaBoards.kt
// PacaBoards.kt
//
// Kotlin port of CircleAI.Workflows/PacaBoards.cs.
//
// (3.3.0) Sprintboard surface ported from paca: rich JSON description, custom
// fields, story points, importance, parent/child relations, status columns with
// position-ordered workflow, drag-and-drop status transitions, sprints with
// lifecycle states, Scrumban swimlanes, per-view persistent configs (filters +
// sort + visible fields), tags, lazy-load pagination per column.

package com.bhengubv.circleai.workflows

import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

/** (3.3.0) Sprint lifecycle. */
enum class SprintState { Planning, Active, Completed }

/** (3.3.0) Status column in the workflow. */
data class StatusColumn(
    val name: String,       // "todo" / "in_progress" / "in_review" / "done"
    val category: String,   // "open" / "in-flight" / "review" / "closed" / "cancelled" / "blocked"
    val position: Int,
    val collapsed: Boolean,
)

/** (3.3.0) Sprint. */
data class PacaSprint(
    val id: String,
    val projectId: String,
    val name: String,
    val goal: String,
    val startDate: Instant,
    val endDate: Instant,
    val state: SprintState,
)

/** (3.3.0) Extra board-only metadata on top of [PacaTask]. */
data class TaskBoardMetadata(
    val projectId: String,
    val number: Int,
    val storyPoints: Int,
    val importance: Int,           // 0..5
    val assigneeMemberId: String?,
    val reporterMemberId: String?,
    val parentTaskNumber: Int?,
    val sprintId: String?,
    val tags: List<String>,
    val customFields: Map<String, String>,
    val positionInColumn: Int,
)

/** (3.3.0) A per-user / per-board "named view". */
data class BoardView(
    val name: String,
    val filterTagsCsv: String?,
    val filterAssignee: String?,
    val sortBy: String?,           // "importance" / "story_points" / "newest"
    val sortDescending: Boolean,
    val visibleColumns: List<String>,
    val visibleFields: List<String>,
)

/**
 * (3.3.0) Board service over a project. Sprints + columns + per-task metadata +
 * views.
 */
class PacaBoard(
    private val tasks: InMemoryPacaStore,
    private val clock: () -> Instant = { Instant.now() },
) {
    private val columns = ConcurrentHashMap<String, StatusColumn>()
    private val sprints = ConcurrentHashMap<String, PacaSprint>()
    private val metadata = ConcurrentHashMap<Pair<String, Int>, TaskBoardMetadata>()
    private val views = ConcurrentHashMap<String, BoardView>()

    init {
        addDefaultColumns()
    }

    private fun addDefaultColumns() {
        columns["todo"] = StatusColumn("todo", "open", 0, false)
        columns["in_progress"] = StatusColumn("in_progress", "in-flight", 1, false)
        columns["in_review"] = StatusColumn("in_review", "review", 2, false)
        columns["done"] = StatusColumn("done", "closed", 3, false)
        columns["cancelled"] = StatusColumn("cancelled", "cancelled", 4, false)
        columns["blocked"] = StatusColumn("blocked", "blocked", 5, true)
    }

    val columnsList: List<StatusColumn> get() = columns.values.sortedBy { it.position }

    fun addColumn(col: StatusColumn) {
        columns[col.name] = col
    }

    fun collapseColumn(name: String, collapsed: Boolean) {
        val col = columns[name] ?: return
        columns[name] = col.copy(collapsed = collapsed)
    }

    /** (3.3.0) Move a task between status columns, updating its in-column position. */
    fun moveTask(projectId: String, number: Int, newStatus: String, newPosition: Int) {
        val task = tasks.getTaskByReference(projectId, "$projectId-$number")
            ?: tasks.listTasks(projectId).firstOrNull { it.number == number }
            ?: throw IllegalStateException("Task not found.")
        require(columns.containsKey(newStatus)) { "Unknown status '$newStatus'." }

        tasks.updateTask(task.copy(status = newStatus))
        val meta = getOrCreateMetadata(projectId, number).copy(positionInColumn = newPosition)
        metadata[projectId to number] = meta
    }

    /** (3.3.0) Attach board metadata to an existing task. */
    fun setTaskMetadata(meta: TaskBoardMetadata) {
        metadata[meta.projectId to meta.number] = meta
    }

    fun getTaskMetadata(projectId: String, number: Int): TaskBoardMetadata? =
        metadata[projectId to number]

    /** (3.3.0) Paginated column read for lazy loading. */
    fun tasksInColumn(projectId: String, status: String, skip: Int = 0, take: Int = 50): List<PacaTask> {
        return tasks.listTasks(projectId)
            .filter { it.status == status }
            .sortedBy { getOrCreateMetadata(it.projectId, it.number).positionInColumn }
            .drop(skip).take(take)
    }

    /** (3.3.0) Tasks bucketed by sprint, useful for the Scrumban board. */
    fun tasksInSprint(sprintId: String): List<PacaTask> {
        return metadata.values
            .filter { it.sprintId == sprintId }
            .mapNotNull { m -> tasks.listTasks(m.projectId).firstOrNull { it.number == m.number } }
    }

    /** (3.3.0) Create a sprint in Planning. */
    fun createSprint(id: String, projectId: String, name: String, goal: String, start: Instant, end: Instant): PacaSprint {
        val s = PacaSprint(id, projectId, name, goal, start, end, SprintState.Planning)
        sprints[id] = s
        return s
    }

    fun getSprint(id: String): PacaSprint? = sprints[id]

    fun startSprint(id: String): PacaSprint = transition(id, SprintState.Active)
    fun completeSprint(id: String): PacaSprint = transition(id, SprintState.Completed)

    private fun transition(id: String, to: SprintState): PacaSprint {
        val sprint = sprints[id] ?: throw IllegalStateException("Sprint '$id' not found.")
        val updated = sprint.copy(state = to)
        sprints[id] = updated
        return updated
    }

    /** (3.3.0) Save a named view (filters + sort + visible fields). */
    fun saveView(view: BoardView) {
        views[view.name] = view
    }

    fun getView(name: String): BoardView? = views[name]

    fun listViews(): List<BoardView> = views.values.sortedBy { it.name }

    private fun getOrCreateMetadata(projectId: String, number: Int): TaskBoardMetadata =
        metadata.computeIfAbsent(projectId to number) {
            TaskBoardMetadata(projectId, number, 0, 3, null, null, null, null, emptyList(), emptyMap(), 0)
        }
}
